diff --git a/Tools/CMake/AMReXInstallHelpers.cmake b/Tools/CMake/AMReXInstallHelpers.cmake
index 1234567..abcdefg 100644
--- a/Tools/CMake/AMReXInstallHelpers.cmake
+++ b/Tools/CMake/AMReXInstallHelpers.cmake
@@ -157,7 +157,7 @@ macro( add_test_install_target _dir _amrex_root )

    # Configure and Build
    # A registered POST_BUILD target in Tests/CMakeTestInstall will then also run the test.
-   add_custom_target(test_install
+   add_custom_target(amrex_test_install
       COMMAND ${CMAKE_COMMAND} -E echo ""
       COMMAND ${CMAKE_COMMAND} -E echo "------------------------------------"
       COMMAND ${CMAKE_COMMAND} -E echo "     Testing AMReX installation     "
